modules/phase_field/src/materials/EpsAndMAndLAnisoGauss.C

```diff
@@ -0,0 +1,931 @@
+
+  for (unsigned int h = 0; h < _grain_num; ++h)
+  {
+    for (unsigned int s = 0; s < _op_num; ++s)
+    {
+      if (h == op_to_grains[s])
+      {
+        _qwg[_qp] = qwR[h];
+        _qxg[_qp] = qxR[h];
+        _qyg[_qp] = qyR[h];
+        _qzg[_qp] = qzR[h];
+      }
+    }
+  }
+
+  //**************************************************************************
+
+  ADReal sumeps = 0.0;
+  ADReal sumepsx = 0.0;
+  ADReal sumepsy = 0.0;
+  ADReal sumepsz = 0.0;
+
+  ADReal sumL = 0.0;
+
+  ADReal sumepsxplus = 0.0;
+  ADReal sumepsyplus = 0.0;
+  ADReal sumepszplus = 0.0;
+
+  ADReal aniso = 0.0;
+  ADReal anisox = 0.0;
+  ADReal anisoy = 0.0;
+  ADReal anisoz = 0.0;
+
+  ADReal anisoxplus = 0.0;
+  ADReal anisoyplus = 0.0;
+  ADReal anisozplus = 0.0;
+
+  ADReal anisoL = 0.0;
+
+  ADReal Val = 0.0;
+  ADReal Val1 = 0.0;
+  ADReal sumval = 0.0;
+
+  ADReal f0saddle = 0.0;
+  ADReal g_gamma = 0.0;
+
+  ADReal eps_base = 0.0;
+  ADReal TotGaussValue = 0.0;
+  ADReal SumTotGaussValue = 0.0;
+  ADReal Totepsbar = 0.0;
+
+  ADReal eps_min = 0.0;
+  ADReal Totepsmin = 0.0;
+
+  f0saddle = 0.124961;
+
+  g_gamma = 0.471404;
+
+  // eps_base = ( (_sigmaBASE *_JtoeV * _length_scale * _length_scale) * _lgb *
+  // (std::sqrt(f0saddle)) ) / (g_gamma) ;
+  eps_base = (3.0 / 4.0) * (_sigmaBASE * _JtoeV * _length_scale * _length_scale) * _lgb;
+
+  //********************** sum (etai*etaj)^2 at each quadrature point **************************
+
+  for (unsigned int m = 0; m < _op_num - 1; ++m)
+  {
+    for (unsigned int n = m + 1; n < _op_num; ++n) // m<n
+    {
+      Val1 = (((*_vals[m])[_qp]) * ((*_vals[m])[_qp])) * (((*_vals[n])[_qp]) * ((*_vals[n])[_qp]));
+      sumval += Val1;
+    }
+  }
+
+  //********************** Main anisotopy calculation **************
+
+  for (unsigned int b = 0; b < _grain_num - 1; ++b)
+  {
+
+    for (unsigned int m = 0; m < _op_num - 1; ++m)
+    {
+
+      for (unsigned int a = b + 1; a < _grain_num; ++a)
+      {
+
+        for (unsigned int n = m + 1; n < _op_num; ++n)
+        {
+
+          if (b == op_to_grains[m])
+          {
+
+            if (a == op_to_grains[n])
+            {
+
+              aniso = eps_base;
+              anisox = 0.0;
+              anisoy = 0.0;
+              anisoz = 0.0;
+
+              anisoxplus = 0.0;
+              anisoyplus = 0.0;
+              anisozplus = 0.0;
+
+              anisoL = _L_BASE;
+
+              TotGaussValue = 0.0;
+
+              // calculate the continuous anisotropy part 1
+
+              Val = (((*_vals[m])[_qp]) * ((*_vals[m])[_qp])) *
+                    (((*_vals[n])[_qp]) * ((*_vals[n])[_qp]));
+
+              // vector of gaussian exponent
+
+              ADReal vsmalbax = (*_grad_vals[m])[_qp](0) - (*_grad_vals[n])[_qp](0);
+              ADReal vsmalbay = (*_grad_vals[m])[_qp](1) - (*_grad_vals[n])[_qp](1);
+              ADReal vsmalbaz = (*_grad_vals[m])[_qp](2) - (*_grad_vals[n])[_qp](2);
```

Comment:
then use a vector variable `RealVectorValue`. This is a general pattern in your code. Please, try to avoid explicit componentwise operations wherever you can!

Reply:
See updated files.
I changed all the ones that could be changed. The ones that are left are those using the four quaternions components like final_gaussian_direction. Those need to have componentwise operations.